src/devices/sound/stt_sa1.cpp

```diff
@@ -0,0 +1,210 @@
+
+	return get_u16be(&m_ram[offset * 2]);
+}
+
+void stt_sa1_device::write(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	if (offset < 0x100 / 2)
+	{
+		const int v = offset >> 4;
+		const int reg = offset & 0xf;
+
+		m_regs[offset] = data;
+
+		switch (reg)
+		{
+			case 1:
+				m_voice[v].addr_start = m_voice[v].addr_cur = (m_voice[v].addr_cur & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: start = %08llx\n", v, m_voice[v].addr_cur >> 12);
+				break;
+			case 2:
+				m_voice[v].addr_start = m_voice[v].addr_cur = (m_voice[v].addr_cur & 0x0000ffff000) | (uint64_t(data & 0x7fff) << 28);
+				m_voice[v].is_looped = BIT(data, 15) != 0;
+				LOG("voice %d: start = %08llx, is_looped = %d\n", v, m_voice[v].addr_cur >> 12, m_voice[v].is_looped);
+				break;
+
+			case 3:
+				m_voice[v].freq = data;
+				LOG("voice %d: step = %08x\n", v, m_voice[v].freq);
+				break;
+
+			case 4:
+				m_voice[v].addr_start = (m_voice[v].addr_start & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: addr_start = %08llx\n", v, m_voice[v].addr_start >> 12);
+				break;
+			case 5:
+				m_voice[v].addr_start = (m_voice[v].addr_start & 0x0000ffff000) | (uint64_t(data) << 28);
+				LOG("voice %d: addr_start = %08llx\n", v, m_voice[v].addr_start >> 12);
+				break;
+
+			case 6:
+				m_voice[v].addr_end = (m_voice[v].addr_end & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: addr_end = %08llx\n", v, m_voice[v].addr_end >> 12);
+				break;
+			case 7:
+				m_voice[v].addr_end = (m_voice[v].addr_end & 0x0000ffff000) | (uint64_t(data) << 28);
+				LOG("voice %d: addr_end = %08llx\n", v, m_voice[v].addr_end >> 12);
+				break;
+
+			case 0x0b:
+				m_voice[v].vol_l = data;
+				LOG("voice %d: vol_l = %08x\n", v, m_voice[v].vol_l);
+				break;
+			case 0x0c:
+				m_voice[v].vol_r = data;
+				LOG("voice %d: vol_r = %08x\n", v, m_voice[v].vol_r);
+				break;
+
+			default:
+				LOG("Unknown register usage: voice %d, register %x, data %04x\n", v, reg, data);
+				break;
+		}
+	}
+	else
+	{
+		put_u16be(&m_ram[offset * 2], data);
+	}
+}
+
+uint16_t stt_sa1_device::key_r()
+{
+	if (!m_enabled)
+		return 0;
+
+	m_stream->update();
+	return m_keyctrl;
+}
+
+void stt_sa1_device::key_w(uint16_t data)
+{
+	if (!m_enabled)
+		return;
+
+	const u16 prev = m_keyctrl;
+
+	m_stream->update();
+
+	m_keyctrl = data;
+
+	for (int v = 0; v < 8; v++) {
+		if (BIT(m_keyctrl, v) && !BIT(prev, v)) {
+			// keyon
+			m_voice[v].enabled = true;
+			m_voice[v].addr_cur = m_voice[v].addr_start;
+		} else if (!BIT(m_keyctrl, v) && BIT(prev, v)) {
+			// keyoff
+			m_voice[v].enabled = false;
+		}
+	}
+}
+
+int8_t stt_sa1_device::read_sample(offs_t offset)
+{
+	if (offset >= 0x600000)
+		return m_readsample_cb(offset); // Held externally in flash chips
+
+	return m_ram[offset];
+}
+
+void stt_sa1_device::device_start()
+{
+	m_stream = stream_alloc(0, 2, clock() / 448);
+	m_ram = make_unique_clear<uint8_t[]>(0x600000); // 12 512k x 8 SRAM chips
```

Comment:
Is the RAM logically words or bytes?

Alternatively, you could store the RAM as a `std::unique_ptr<uint16_t []>`:
* No need to double the offset and use `get_u16be`/`put_u16be` when reading/writing.
* Use `util::big_endian_cast<uint8_t>(m_ram.get())[offset]` (from endianness.h) in `read_sample`.

Reply:
The registers are 16-bit but the RAM chips themselves are 8-bit (`U31, U32, U33, U37, U38, U39, U40, U41, U44, U45, U47, U48 (HY628400): Toshiba TC554001, 512k x 8 SRAM`) which is why I implemented the RAM as `uint8_t` in the device here just to be consistent with the real hardware.

Would you consider using `uint16_t` for the RAM the better choice here, performance and such considered? It's all implemented in the FPGA and I can't verify anything beyond how the code uses the chip so I don't have a preference here as long as it's works.

Reply:
Oh dammit.  I see what I’m missing.

It isn’t that the RAM and Flash is behind the “sound chip” per se.  It’s just that it has an external 16-bit memory bus that’s attached to the RAM and Flash, and the the FPGA also allows the CPU to access a window into this space.

The “sound chip” device should only be handling its own register reads and writes, and it should derive from `device_rom_interface` to access its sample storage.  The driver can install the RAM and Flash in its address space.

Are you comfortable looking for examples of `device_rom_interface` and doing it, or do I need to have a crack at it in the next few days?

Reply:
Oh, and just to be explicit, the 8-bit RAMs can easily be wired onto a 16-bit bus, you just wire 6 of them to the upper byte and the other 6 to the lower byte.  Lane select is easy for byte writes because you can enable one or both of the chips.

Reply:
I gave it a shot. Everything seems to be working still. Checked endianness of both the RAM and data being read from flash and everything is looking how it should still.

https://github.com/mamedev/mame/pull/11929/commits/e142229da1507ca2af5cbde9f924447684318a4e

Reply:
I’ll do a bit of cleanup and push to this PR branch so you can re-test it.